pkg/keylock/keylock.go

```diff
@@ -83,11 +92,35 @@ func (kl *keyLock) Acquire(id string) LockHandle {
 	return *h
 }
 
+func (kl *keyLock) AcquireWithTimeout(id string, duration time.Duration) *LockHandle {
+	h := kl.getOrCreateLock(id)
+	timeout := time.After(duration)
+	for {
+		select {
+		case <-timeout:
+			h.refcnt--
+			return nil
+		default:
+			if tryLock(h) {
+				h.genNum++
+				return h
+			}
+			time.Sleep(1 * time.Millisecond)
+		}
+	}
+}
+
+func tryLock(h *LockHandle) bool {
+	return atomic.CompareAndSwapInt32((*int32)(unsafe.Pointer(h.mutex)), 0, mutexLocked)
+}
+
 func (kl *keyLock) Release(h *LockHandle) error {
+	if h == nil {
```

Comment:
nit: combine with the next if. `if h== nil || len(h.id) == 0`

Reply:
done